dwave/cloud/api/client.py

```diff
@@ -131,9 +133,108 @@ def request(self, method: str, *args, **kwargs):
         return response
 
 
-class VersionedAPISession(LoggingSession):
+class PayloadCompressingSession(LoggingSession):
+    """A :class:`requests.Session` subclass (technically, a further specialized
+    :class:`.LoggingSession`) that adds support for payload compression on the
+    fly.
+
+    Args:
+        compress (bool, default=False):
+            Preemptively compress HTTP request body, even on PUT and POST requests.
+            When compression is enabled, the payload is always compressed in 32KiB
+            chunks and streamed to minimize memory overhead.
+
+    .. note::
+        With compression enabled, the request body is transferred using chunked
+        encoding (``Transfer-Encoding: chunked``) and content is compressed using
+        deflate compression algorithm (``Content-Encoding: deflate``).
+
+    .. versionadded:: 0.13.3
+        Preemptive payload compression support added to :class:`.DWaveAPIClient`.
+
+    """
+
+    _compress: bool = False
+
+    def __init__(self, compress: bool = False, **kwargs):
+        self._compress = compress
+        super().__init__(**kwargs)
+
+    def set_payload_compress(self, compress: bool = True) -> bool:
+        """Modify payload compression on upload for current session.
+
+        Args:
+            compress (bool, default=True):
+                Enable request body compression with ``deflate`` on upload
+                requests like ``POST`` and ``PUT``. Make sure the server supports
+                compression on inbound requests, as most server don't.
+
+        Returns:
+            bool:
+                Previous compression setting.
+        """
+        previous = self._compress
+        self._compress = compress
+        return previous
+
+    @staticmethod
+    def _iter_compressed(data: Union[bytes, IO, Iterable[bytes]],
+                         chunk_size: Optional[int] = None):
+        if chunk_size is None:
+            # match zlib window size
+            chunk_size = 2**15
+
+        if isinstance(data, bytes):
+            data = io.BytesIO(data)
+
+        if hasattr(data, 'read'):
+            def get_chunks(f, chunk_size):
+                while chunk := f.read(chunk_size):
+                    yield chunk
+            chunks = get_chunks(data, chunk_size)
+
+        else:
+            # assume data is iterable
+            chunks = data
+
+        zbuf = zlib.compressobj()
```

Comment:
I believe this uses default compression. I have no reason to think that's a bad level, but do we want to make the configurable? Might be too low level of a knob to expose :thinking:

Reply:
Fair point; yes, we use the default compression level (`-1` / `Z_DEFAULT_COMPRESSION`).
>  `Z_DEFAULT_COMPRESSION` represents a default compromise between speed and compression (currently equivalent to level 6).

It's a very low-level knob to expose (on an already pretty low-level interface), and I can't think of a practical use case for this feature. However, if anyone asks for it, it's trivial to add it.. so, until then..

Reply:
Thinking about this, making the compression algorithm selectable (adding Brotli or Zstd) would be nice, but probably an overkill with questionable benefits.

Reply:
Did some zlib level benchmarks on a random P16 problem (float biases between -1 and 1). Uncompressed encoded QPU data is 492039 bytes. And levels 5-9 result in the same compression ratio:
```
0: 492090
1: 376840
2: 376748
3: 376739
4: 370557
5: 370552
6: 370552
7: 370552
8: 370552
9: 370552
```

Reply:
Totally agree it's too low level to add eagerly